src/advanced-concepts/async-programming.md

```diff
@@ -1 +1,414 @@
+
+A call to a shared function of any actor A, whether from an [_external client_](/internet-computer-programming-concepts/actors/canister-calling.html), [from itself](#async-and-await) or from another actor B (as an [Inter-Canister Call](#inter-canister-calls)), results in an _incoming message_ to actor A.
+
+A single message is executed _atomically_. This means that the code executed within one message either executes successfully or not at all. This also means that any _state mutations_ within a single message are either all committed or none of them are committed.
+
+**An `await` ends the current message and splits the execution of a function into [separate messages](#shared-functions-that-await).**
+
+### Atomic functions
+
+An atomic function is one that executes within one single message. The function either executes successfully or has no effect at all. If an atomic function fails, we know for sure its state mutations have not been committed.
+
+**If a shared function does not `await` in its body, then it is executed atomically.**
+
+```motoko
+{{#include _mo/async-calls4.mo:a}}
+```
+
+Our function `atomic` mutates the state, performs a computation and mutates the `state` again. Both the computation and the state mutations belong to the same message execution. The whole function either succeeds and commits all state mutations or it fails and does not commit any changes at all.
+
+The second function `atomic_fail` is another atomic function. It also performs a computation and state mutations within one single message. **But this time, the computation [traps](#traps) and both state mutations are not committed, even though the trap happened after the first state mutation.**
+
+Unless an [`Error`](#errors) is thrown _intentionally_ during execution, the order in which computations and state mutations occur is not relevant for the atomicity of a function. The whole function is executed as a single message that either fails or succeeds in its entirety.
+
+### Shared functions that `await`
+
+The [async-await](#async-and-await) example earlier looks simple, but there is a lot going on there. The function call is executed as several separate messages:
+
+The first line `let future : async Nat = read()` is executed as part of the first message.
+
+The second line `let result = await future;` keyword ends the first message and any state changes made during execution of the message are committed. The `await` also calls `read()` and suspends execution until a response is received.
+
+The call to `read()` is executed as a separate message and could possibly be executed remotely in another actor, see [inter-canister calls](#inter-canister-calls). The message sent to `read()` could possibly result in several messages if the `read()` function also `await`s in its body.
+
+If the sent message executes successfully, a _callback_ is made to `call_read` that is executed as yet another message as the the last line `result + 1`.
```

Comment:
```suggestion
If the sent message executes successfully, a _callback_ is made to `call_read` that is executed as yet another message as the last line `result + 1`.
```

Reply:
The callback writes the result into the future, completing it, and resumes the code that was waiting on the future, i.e. the last line `result +1`.